optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc

```diff
@@ -585,6 +610,77 @@ Joining should only be used in cases where the other fact is actually required f
 the stream.
 ====
 
+[[constraintStreamsMappingTuples]]
+=== Mapping tuples
+
+Mapping enables you to transform each tuple in a constraint stream by applying a mapping function to it.
+The result of such mapping is `UniConstraintStream` of the mapped tuples.
+
+[source,java,options="nowrap"]
+----
+    private Constraint computerWithBiggestProcess(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(CloudProcess.class) // UniConstraintStream<CloudProcess>
+                .map(CloudProcess::getComputer)           // UniConstraintStream<CloudComputer>
+                ...
+    }
+----
+
+[NOTE]
+====
+In the example above, the mapping function produces duplicate tuples if two different `CloudProcess`es share a single `CloudComputer`.
+That is, such `CloudComputer` will appear in the resulting constraint stream twice.
+See <<constraintStreamsDistinct,`distinct()`>> for how to deal with duplicate tuples.
+====
+
+[[constraintStreamsMappingFunction]]
+==== Designing the mapping function
+
+When designing the mapping function, we recommend you follow these guidelines for optimal performance:
+
+* Keep the function pure.
+  The mapping function should only depend on its input.
+  That is, given the same input, it always returns the same output.
+* Keep the function simple.
+  The function should only take the argument and return another simple value,
+  without performing any significant computation.
+  Not following this recommendation may result in performance loss,
+  as the function may be called frequently.
+* Keep the function bijective.
+  No two input tuples should map to the same output tuple, or to tuples that are equal.
+  Not following this recommendation will create a constraint stream with duplicate tuples,
+  and may force you to use <<constraintStreamsDistinct,`distinct()`>> later.
+* Use immutable data carriers.
+  The tuples returned by the mapping function should be immutable and identified by their contents and nothing else.
+  If two tuples carry objects which equal one another,
+  those two tuples should likewise equal and preferably be the same instance.
+  https://openjdk.java.net/jeps/395[Java's Records] are an ideal tool for the purpose.
+
+[[constraintStreamsDistinct]]
+==== Dealing with duplicate tuples
```

Comment:
I'd love to see the word "distinct" in this section header as that's the operation it's talking about.
In case I use it (or I inherit a piece of code that uses distinct()) and I want to fully understand what it does by reading the docs, so I need to find it in the TOC.